### PR TITLE
8304913: Use OperatingSystem, Architecture, and Version in jlink

### PR DESCRIPTION
**Pull request by RogerRiggs**
Refactor the Platform class of jlink to use jdk.internal.util OperatingSystem and Architecture instead of os.name and os.arch. 
They are direct replacements for the Platform enums except for UNKNOWN; its use is refactored to report errors via exceptions.

Neither os.name nor os.arch should be assumed to be changeable; 
one test case is removed because it assumes os.name can be changed on the command line.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8304913](https://bugs.openjdk.org/browse/JDK-8304913): Use OperatingSystem, Architecture, and Version in jlink

### Reviewers
 * [Jaikiran Pai](https://openjdk.org/census#jpai) (@jaikiran - **Reviewer**) ⚠️ Review applies to [cfc404c1](https://git.openjdk.org/jdk/pull/13585/files/cfc404c1d0d9efba127b9ef675b203dfd075a600)
 * [Alan Bateman](https://openjdk.org/census#alanb) (@AlanBateman - **Reviewer**) ⚠️ Review applies to [cfc404c1](https://git.openjdk.org/jdk/pull/13585/files/cfc404c1d0d9efba127b9ef675b203dfd075a600)
 * [Mandy Chung](https://openjdk.org/census#mchung) (@mlchung - **Reviewer**) ⚠️ Review applies to [cfc404c1](https://git.openjdk.org/jdk/pull/13585/files/cfc404c1d0d9efba127b9ef675b203dfd075a600)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/13585/head:pull/13585` \
`$ git checkout pull/13585`

Update a local copy of the PR: \
`$ git checkout pull/13585` \
`$ git pull https://git.openjdk.org/jdk.git pull/13585/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 13585`

View PR using the GUI difftool: \
`$ git pr show -t 13585`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/13585.diff">https://git.openjdk.org/jdk/pull/13585.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/13585#issuecomment-1533664130)

### CODE REVIEW

File: src/jdk.jlink/share/classes/jdk/tools/jlink/builder/DefaultImageBuilder.java
Comment (jaikiran) [inline src/jdk.jlink/share/classes/jdk/tools/jlink/builder/DefaultImageBuilder.java:575]: Hello Roger, I couldn't find any usages of this (existing) `public static ExecutableImage getExecutableImage(Path root)` in the code. Maybe we could remove it?

Comment (jaikiran) [reply src/jdk.jlink/share/classes/jdk/tools/jlink/builder/DefaultImageBuilder.java:575]: If we do remove this method, the `private static Set<String> retrieveModules(Path root)` method in this same class will become unused too, so can be removed too.

File: src/jdk.jlink/share/classes/jdk/tools/jlink/internal/Platform.java
Comment (jaikiran) [inline src/jdk.jlink/share/classes/jdk/tools/jlink/internal/Platform.java:64]: Nit - the enclosing `()` aren't needed.

File: src/jdk.jlink/share/classes/jdk/tools/jlink/internal/Platform.java
Comment (jaikiran) [inline src/jdk.jlink/share/classes/jdk/tools/jlink/internal/Platform.java:71]: Should we create a single `private static final Platform`, in this class, to represent the current platform, and keep returning it instead of creating a new instance on each call?

File: src/jdk.jlink/share/classes/jdk/tools/jlink/builder/ImageBuilder.java
Comment (jaikiran) [inline src/jdk.jlink/share/classes/jdk/tools/jlink/builder/ImageBuilder.java:86]: I think it is OK to change this implementation to now throw this exception, since this is an internal interface (from what I can see in the module definition).

**Review by jaikiran — APPROVED** [review]
These latest changes look OK to me.

I have very limited knowledge of this area, so it would be good to wait for another review.

File: src/jdk.jlink/share/classes/jdk/tools/jlink/builder/DefaultImageBuilder.java
Comment (AlanBateman) [inline src/jdk.jlink/share/classes/jdk/tools/jlink/builder/DefaultImageBuilder.java:180]: What is "ile", can we rename it to "iae"?

File: src/jdk.jlink/share/classes/jdk/tools/jlink/internal/plugins/ExcludeVMPlugin.java
Comment (AlanBateman) [inline src/jdk.jlink/share/classes/jdk/tools/jlink/internal/plugins/ExcludeVMPlugin.java:256]: Confusing to name it "ile" here too.

File: src/jdk.jlink/share/classes/jdk/tools/jlink/internal/plugins/ExcludeVMPlugin.java
Comment (AlanBateman) [inline src/jdk.jlink/share/classes/jdk/tools/jlink/internal/plugins/ExcludeVMPlugin.java:254]: If you line up the -> then it might be clearer to the eyes.

File: test/jdk/tools/jlink/plugins/CDSPluginTest.java
Comment (AlanBateman) [inline test/jdk/tools/jlink/plugins/CDSPluginTest.java:102]: Good to see this part of the test going away.

**Review by jaikiran — APPROVED** [review]

File: src/jdk.jlink/share/classes/jdk/tools/jlink/internal/Platform.java
Comment (AlanBateman) [inline src/jdk.jlink/share/classes/jdk/tools/jlink/internal/Platform.java:53]: "Jmod architecture" is confusing here, it's the architecture component of the ModuleTarget attribute.

**Review by AlanBateman — APPROVED** [review]
Aside from one confusing comment, this change looks okay.

**Review by mlchung — APPROVED** [review]
I suggest to remove `Platform::is64Bit` and simply call `Architecture::is64bit`.  Otherwise, looks fine.

File: src/jdk.jlink/share/classes/jdk/tools/jlink/internal/Platform.java
Comment (mlchung) [inline src/jdk.jlink/share/classes/jdk/tools/jlink/internal/Platform.java:64]: `Platform::is64Bit` tests if the target architecture is 64-bit but `Architecture.is64bit()` tests the runtime architecture.

File: src/jdk.jlink/share/classes/jdk/tools/jlink/internal/Platform.java
Comment (mlchung) [inline src/jdk.jlink/share/classes/jdk/tools/jlink/internal/Platform.java:65]: Have you considered having an instance `Architecture::is64bit` method in lieu of the current static method?

Comment (Glavo) [inline src/jdk.jlink/share/classes/jdk/tools/jlink/internal/Platform.java:65]: This change causes incorrect results to be returned on Linux ARM32.

Comment (RogerRiggs) [reply src/jdk.jlink/share/classes/jdk/tools/jlink/internal/Platform.java]: Possibly, if a need subsequently arises. 
The only current use is in CDSPlugin and it is testing if the current/runtime platform is 64-bit so it can create the non-oop archive. The `Architecture.is64bit()` provides that.  The current `Platform.is64Bit()` method is not needed.

Comment (RogerRiggs) [reply src/jdk.jlink/share/classes/jdk/tools/jlink/internal/Platform.java]: My mistake; see above, `Platform.is64Bit()` is not needed by CDSPlugin.

**Pull request closed without merging**
